replica/table.cc

```diff
@@ -634,11 +635,25 @@ table::update_cache(compaction_group& cg, lw_shared_ptr<memtable> m, std::vector
         m->mark_flushed(std::move(new_ssts_ms));
         try_trigger_compaction(cg);
     });
+    std::exception_ptr ex;
     if (cache_enabled()) {
+      try {
         co_return co_await _cache.update(std::move(adder), *m);
-    } else {
-        co_return co_await _cache.invalidate(std::move(adder)).then([m] { return m->clear_gently(); });
+      } catch (...) {
+        ex = std::current_exception();
+      }
+    }
+    co_await _cache.invalidate(std::move(adder));
+    co_await m->clear_gently();
+    if (ex) {
+        // Just report the exception as invalidating the cache is fine as fallback.
+        tlogger.warn("Failed to update the row cache: {}", std::move(ex));
     }
+  } catch (...) {
+    std::invoke([&] () noexcept {
+        on_fatal_internal_error(tlogger, fmt::format("failed to update the row cache: {}", std::current_exception()));
```

Comment:
Is `table::update_cache` the appropriate place to turn exceptions into `abort`? (I'm not suggesting it isn't — I just think it would be good to have some philosophy about which places are allowed to throw/fail and which have to abort() instead).

If any error in the body of `table::update_cache` terminates the program, perhaps we should make `table_update_cache` infallible (currently it still can fail when allocating the coroutine frame) and document that, so that callers don't have to anticipate exceptions coming from it.

Reply:
I think that allowing update cache to fail on bad_alloc and in response invalidating parts of the cache (and freeing up memory in the process) makes a lot of sense.
@tgrabiec's what do you think.

Reply:
But invalidate() itself can fail in the same way, no?

This would be very weakly tested, better to abort.

Reply:
Ok, I see you implemented it. But what if invalidate() fails? Ah, it would be caught later.

Isn't adder being used-after-move?

Isn't invalidate() called unconditionally?

Where's the test for this?

Reply:
> Isn't adder being used-after-move?

Good point.

> Isn't invalidate() called unconditionally?

No, since the `co_return co_await _cache.update()` above.

> Where's the test for this?

We need to add one that will inject a failure in update